### ISSUE TITLE
 Error in `/opt/TurboVNC/bin/Xvnc': free(): invalid pointer: 0x00000000048b6310

### ISSUE THREAD

**Issue by rangeshgupta-altair** [TurboVNC/turbovnc#189]
Below is the XVNC log. It is happening randomly. So not sure the exact steps to reproduce this issue. We are using TurboVNC 2.0

**XVNC log:**

Segmentation fault at address 0x6cf700c
 
Fatal server error:
Caught signal 11 (Segmentation fault). Server aborting
 
*** Error in `/opt/TurboVNC/bin/Xvnc': free(): invalid pointer: 0x00000000048b6310 ***
======= Backtrace: =========
/lib64/libc.so.6(+0x81429)[0x2b56f97df429]
/opt/TurboVNC/bin/Xvnc[0x4a1f4e]
/opt/TurboVNC/bin/Xvnc[0x4fd7c0]
/opt/TurboVNC/bin/Xvnc(FreeGC+0x3b)[0x4529fb]
/opt/TurboVNC/bin/Xvnc(rfbDCDeviceCleanup+0xad)[0x4a0dfd]
/opt/TurboVNC/bin/Xvnc[0x487533]
/opt/TurboVNC/bin/Xvnc[0x4ee33d]
/opt/TurboVNC/bin/Xvnc[0x435b45]
/opt/TurboVNC/bin/Xvnc(CloseDownDevices+0x125)[0x4362f5]
/opt/TurboVNC/bin/Xvnc(AbortServer+0x1d)[0x52719d]
/opt/TurboVNC/bin/Xvnc[0x5273a5]
/opt/TurboVNC/bin/Xvnc[0x528dcb]
/lib64/libpthread.so.0(+0xf680)[0x2b56f9342680]
/opt/TurboVNC/bin/Xvnc(fbBlt+0xa33)[0x4ae693]
/opt/TurboVNC/bin/Xvnc(fbCopyNtoN+0x21d)[0x4bf33d]
/opt/TurboVNC/bin/Xvnc(miCopyRegion+0xca)[0x4e787a]
/opt/TurboVNC/bin/Xvnc(miDoCopy+0x45c)[0x4e7ebc]
/opt/TurboVNC/bin/Xvnc(fbCopyArea+0x4c)[0x4bf04c]
/opt/TurboVNC/bin/Xvnc[0x4ffd6f]
/opt/TurboVNC/bin/Xvnc(ProcCopyArea+0x148)[0x43b7a8]
/opt/TurboVNC/bin/Xvnc(Dispatch+0x4b1)[0x43f6b1]
/opt/TurboVNC/bin/Xvnc(main+0x2da)[0x432eba]
/lib64/libc.so.6(__libc_start_main+0xf5)[0x2b56f97803d5]
/opt/TurboVNC/bin/Xvnc[0x432b29]
======= Memory map: ========
00400000-0076e000 r-xp 00000000 fd:00 539308667                          /opt/TurboVNC/bin/Xvnc
0096d000-00976000 rw-p 0036d000 fd:00 539308667                          /opt/TurboVNC/bin/Xvnc
00976000-009ad000 rw-p 00000000 00:00 0 
01e55000-06cf7000 rw-p 00000000 00:00 0                                  [heap]
2b56f8e0d000-2b56f8e2f000 r-xp 00000000 fd:00 44994                      /usr/lib64/ld-2.17.so
2b56f8e2f000-2b56f8e34000 rw-p 00000000 00:00 0 
2b56f8e4c000-2b56f8edb000 rw-p 00000000 00:00 0 
2b56f8edb000-2b56f8f5b000 rw-s 00000000 00:04 1002602535                 /SYSV00000000 (deleted)
2b56f8f5b000-2b56f8fdb000 rw-s 00000000 00:04 1002700840                 /SYSV00000000 (deleted)
2b56f902e000-2b56f902f000 r--p 00021000 fd:00 44994                      /usr/lib64/ld-2.17.so
2b56f902f000-2b56f9030000 rw-p 00022000 fd:00 44994                      /usr/lib64/ld-2.17.so
2b56f9030000-2b56f9031000 rw-p 00000000 00:00 0 
2b56f9031000-2b56f9132000 r-xp 00000000 fd:00 45009                      /usr/lib64/libm-2.17.so
2b56f9132000-2b56f9331000 ---p 00101000 fd:00 45009                      /usr/lib64/libm-2.17.so
2b56f9331000-2b56f9332000 r--p 00100000 fd:00 45009                      /usr/lib64/libm-2.17.so
2b56f9332000-2b56f9333000 rw-p 00101000 fd:00 45009                      /usr/lib64/libm-2.17.so
2b56f9333000-2b56f934a000 r-xp 00000000 fd:00 45027                      /usr/lib64/libpthread-2.17.so
2b56f934a000-2b56f9549000 ---p 00017000 fd:00 45027                      /usr/lib64/libpthread-2.17.so
2b56f9549000-2b56f954a000 r--p 00016000 fd:00 45027                      /usr/lib64/libpthread-2.17.so
2b56f954a000-2b56f954b000 rw-p 00017000 fd:00 45027                      /usr/lib64/libpthread-2.17.so
2b56f954b000-2b56f954f000 rw-p 00000000 00:00 0 
2b56f954f000-2b56f955c000 r-xp 00000000 fd:00 297301                     /usr/lib64/libpam.so.0.83.1
2b56f955c000-2b56f975c000 ---p 0000d000 fd:00 297301                     /usr/lib64/libpam.so.0.83.1
2b56f975c000-2b56f975d000 r--p 0000d000 fd:00 297301                     /usr/lib64/libpam.so.0.83.1
2b56f975d000-2b56f975e000 rw-p 0000e000 fd:00 297301                     /usr/lib64/libpam.so.0.83.1
2b56f975e000-2b56f9921000 r-xp 00000000 fd:00 45001                      /usr/lib64/libc-2.17.so
2b56f9921000-2b56f9b20000 ---p 001c3000 fd:00 45001                      /usr/lib64/libc-2.17.so
2b56f9b20000-2b56f9b24000 r--p 001c2000 fd:00 45001                      /usr/lib64/libc-2.17.so
2b56f9b24000-2b56f9b26000 rw-p 001c6000 fd:00 45001                      /usr/lib64/libc-2.17.so
2b56f9b26000-2b56f9b2b000 rw-p 00000000 00:00 0 
2b56f9b2b000-2b56f9b48000 r-xp 00000000 fd:00 45396                      /usr/lib64/libaudit.so.1.0.0
2b56f9b48000-2b56f9d48000 ---p 0001d000 fd:00 45396                      /usr/lib64/libaudit.so.1.0.0
2b56f9d48000-2b56f9d49000 r--p 0001d000 fd:00 45396                      /usr/lib64/libaudit.so.1.0.0
2b56f9d49000-2b56f9d4a000 rw-p 0001e000 fd:00 45396                      /usr/lib64/libaudit.so.1.0.0
2b56f9d4a000-2b56f9d54000 rw-p 00000000 00:00 0 
2b56f9d54000-2b56f9d56000 r-xp 00000000 fd:00 45007                      /usr/lib64/libdl-2.17.so
2b56f9d56000-2b56f9f56000 ---p 00002000 fd:00 45007                      /usr/lib64/libdl-2.17.so
2b56f9f56000-2b56f9f57000 r--p 00002000 fd:00 45007                      /usr/lib64/libdl-2.17.so
2b56f9f57000-2b56f9f58000 rw-p 00003000 fd:00 45007                      /usr/lib64/libdl-2.17.so
2b56f9f58000-2b56f9f5c000 r-xp 00000000 fd:00 45392                      /usr/lib64/libcap-ng.so.0.0.0
2b56f9f5c000-2b56fa15c000 ---p 00004000 fd:00 45392                      /usr/lib64/libcap-ng.so.0.0.0
2b56fa15c000-2b56fa15d000 r--p 00004000 fd:00 45392                      /usr/lib64/libcap-ng.so.0.0.0
2b56fa15d000-2b56fa15e000 rw-p 00005000 fd:00 45392                      /usr/lib64/libcap-ng.so.0.0.0
2b56fa15e000-2b56fa94f000 rw-s 00000000 00:04 1002405923                 /SYSV00000000 (deleted)
2b56fa94f000-2b56fb140000 rw-p 00000000 00:00 0 
2b56fb140000-2b56fb931000 rw-s 00000000 00:04 1002733609                 /SYSV00000000 (deleted)
2b56fb931000-2b56fc126000 rw-p 00000000 00:00 0 
2b56fc1e7000-2b56fc5b6000 rw-s 00000000 00:04 1002799147                 /SYSV00000000 (deleted)
2b56fc5b6000-2b56fc985000 rw-s 00000000 00:04 1002831916                 /SYSV00000000 (deleted)
2b56fc985000-2b56fd985000 rw-s 00000000 00:04 1005551714                 /SYSV00000000 (deleted)
2b56fd985000-2b56fdd54000 rw-s 00000000 00:04 1008599143                 /SYSV00000000 (deleted)
2b56fdd54000-2b56fe123000 rw-s 00000000 00:04 1008631920                 /SYSV00000000 (deleted)
2b56fe123000-2b56fe4f2000 rw-s 00000000 00:04 1008730226                 /SYSV00000000 (deleted)
2b56fe4f2000-2b56fe8c1000 rw-s 00000000 00:04 1008762995                 /SYSV00000000 (deleted)
2b56fe8c1000-2b56fe8d6000 r-xp 00000000 fd:00 84                         /usr/lib64/libgcc_s-4.8.5-20150702.so.1
2b56fe8d6000-2b56fead5000 ---p 00015000 fd:00 84                         /usr/lib64/libgcc_s-4.8.5-20150702.so.1
2b56fead5000-2b56fead6000 r--p 00014000 fd:00 84                         /usr/lib64/libgcc_s-4.8.5-20150702.so.1
2b56fead6000-2b56fead7000 rw-p 00015000 fd:00 84                         /usr/lib64/libgcc_s-4.8.5-20150702.so.1
2b5700000000-2b5700021000 rw-p 00000000 00:00 0 
2b5700021000-2b5704000000 ---p 00000000 00:00 0 
7fffe1180000-7fffe11a1000 rw-p 00000000 00:00 0                          [stack]
7fffe11cc000-7fffe11ce000 r-xp 00000000 00:00 0                          [vdso]
ffffffffff600000-ffffffffff601000 r-xp 00000000 00:00 0                  [vsyscall]
XIO:  fatal IO error 11 (Resource temporarily unavailable) on X server ":2"
      after 120 requests (120 known processed) with 0 events remaining.

**Comment by dcommander** [comment]
Let's start with the fact that your server is two generations out of date and you haven't even upgraded to the latest bug-fix release (2.0.2) from that generation.  I see at least three or four bug fixes introduced since 2.0 that may have fixed this, and it's as likely as not that the upgraded X server code we're using now fixed it as well (TurboVNC 2.0 was still using xorg-xserver 1.12.x.)

Referring to https://turbovnc.org/About/ProfessionalServices, support can be provided for obsolete branches of TurboVNC such as 2.0, but that is a paid service.  It would take hours of labor to cherry pick the relevant bug fixes from more recent TurboVNC releases back into the 2.0.x branch, spin builds from that branch, and work with your organization to ensure that the bug is fixed from your point of view.  I can justify using the VirtualGL/TurboVNC General Fund to pay for that labor if the bug exists in the latest release of TurboVNC, because fixing the bug would improve the quality of the product in that case.  However, if it's a bug that was already fixed, then I can't fix it again for free.

In short, your options are:
1. Pay for my labor to diagnose and fix this (contact me through e-mail for rates) and to provide a 2.0.x bug-fix build of the TurboVNC Server for your organization.
2. Diagnose and fix it yourself and submit a PR against the 2.0.x branch.
3. Upgrade to the latest server (I recommend the 2.2.x stable pre-release build from https://turbovnc.org/DeveloperInfo/PreReleases, since that will become 2.2.3 hopefully by the end of this month.)  If the problem still exists there, then I can diagnose and fix it for free.

**Comment by dcommander** [comment]
@rangeshgupta-altair I apologize for not noticing that you are with Altair.  Please contact me through e-mail, as I already have a contractual relationship with your company.

**Comment by dcommander** [comment]
Moving to offline support, since I have a contractual relationship with Altair and this issue likely only affects an obsolete release.
